designs/use-selectors-at-cache.md

```diff
@@ -0,0 +1,124 @@
+Filter cache ListWatch using selectors
+===================
+## Motivation
+
+The controller-runtime subscribe to changes using a cache mechanism that get
+updated using kubernetes informers, those informers implement a `ListWatch`
+that specify the namespace and type specified at controller builder time.
+
+So the only possible filtering of that cache is namespace and resource type,
+this means that if some user is interested only on Reconcile some resource with
+specific fields it will still cache all the instances on that type increasing
+the memory and CPU load. This also get worse in case controller-runtime is
+deployed as a daemonset instead of a deployment since all the nodes with
+daemonset's pods will expect to receive notification of all the instances and
+that can end with scalation problems depending on the type or resource it's
+watching.
+
+As side effect of this proposal the request that does not match the filtered
+will not be returned, so users has to be warned that if this filtering
+is used they know it and they are careful with the request they are doing.
+
+As a side note this proposal sidesteps some of the long standing issues with
+other [solutions](https://github.com/kubernetes-sigs/controller-runtime/issues/244#issuecomment-451336202)
+like problematic interactions between controllers or the need for a complex
+refactorings by adding the possibility for users to define the ListWatch
+selectors but with an API that will be used by users that know what are the
+implications.
+
+The alternative that can be done with the current controller-runtme is
+to implement a custom cache that do the filtering since cache is pluggable
+into the controller-runtime [1]
+
+This proposal is related to the following issue [2]
+
+## Proposal
+
+Add a new selector code at `pkg/cache/selector/selector.go` with common structs
```

Comment:
Why a new `selector` package rather than putting this into the existing `cache` package?

Reply:
`pkg/cache/internal` is importing it, at golang you cannot import "parent" packages.

Reply:
So define it in pkg/cache/internal and alias it from pkg/cache?

Reply:
Make sense, done.